strawberryfields/apps/train/cost.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Submodule for computing gradients, evaluating cost functions, and optimizing GBS
+circuits.
+
+In the context of stochastic optimization, cost functions are expressed as expectation values
+over the GBS distribution. Within the WAW parametrization, gradients of such cost functions can be
+expressed as expectation values over the GBS distribution. This module contains methods for
+calculating these gradients and for using gradient-based methods to optimize GBS circuits. In the
+case of optimization with respect to a Kullback-Leibler divergence or log-likelihood cost
+function, gradients can be computed efficiently, leading to fast training. This module also
+includes the ability to optimize GBS circuits using these efficient methods.
+"""
+
+import numpy as np
+
+
+class KL:
+    r"""Optimization of GBS distributions based on a Kullback-Liebler divergence cost function.
+
+    In a standard unsupervised learning scenario, data are assumed to be sampled from an unknown
+    distribution and a common goal is to learn that distribution. Training of a model
+    distribution can be performed by minimizing the Kullback-Leibler (KL) divergence:
+    .. math::
+
+        KL(P, Q) = -\frac{1}{T}\sum_t \log[P(S^{(t)})]-\log(T),
+
+    where :math:`S^{(t)}` is the *t*-th element in the data, :math:`P(S^{(t)})` is the probability
+    of observing that element when sampling from the model distribution, and :math:`T` is the
+    total number of elements in the data. For the GBS distribution in the WAW parametrization,
+    the gradient of the KL divergence can be written as
+    .. math::
+
+        \partial_\theta KL(\theta) = - \sum_{k=1}^m(\langle n_k\rangle_{\text{data}-
+        \langle n_k\rangle)_{\text{GBS}}\partial_\theta w_k,
+
+    where :math:`\langle n_k\rangle)` denote average photon numbers in mode *k*. This class
+    provides methods to compute gradients and evaluate the cost function.
+
+    **Example usage**
+
+    # TODO: add ex usage for KL class
+
+    Args:
+        data (np.array): Array of samples representing the training data
+        vgbs: Variational GBS class in the WAW parametrization.
+
+    """
+
+    def __init__(self, data, vgbs):
```

Comment:
One thing I just remembered is that we generally add type hinting to the function definitions in the apps layer.
Do we want to follow this in our new GBS variational training feature?

Currently we do have type hinting in https://github.com/XanaduAI/strawberryfields/pull/387 but not in https://github.com/XanaduAI/strawberryfields/pull/388.